// @shopify/polaris/dist/esm/components/ColorPicker/components/AlphaPicker/AlphaPicker.js
import React$1, { PureComponent } from 'react';
import { hsbToRgb } from '../../../../utilities/color-transformers.js';
import styles from '../../ColorPicker.scss.js';
import { Slidable as Slidable$1 } from '../Slidable/Slidable.js';
import { alphaForDraggerY, calculateDraggerY } from './utilities.js';

class AlphaPicker extends PureComponent {
  constructor(...args) {
    super(...args);
    this.state = {
      sliderHeight: 0,
      draggerHeight: 0
    };

    this.setSliderHeight = node => {
      if (node == null) {
        return;
      }

      this.setState({
        sliderHeight: node.clientHeight
      });

      if (process.env.NODE_ENV === 'development') {
        setTimeout(() => {
          this.setState({
            sliderHeight: node.clientHeight
          });
        }, 0);
      }
    };

    this.setDraggerHeight = height => {
      this.setState({
        draggerHeight: height
      });
    };

    this.handleChange = ({
      y
    }) => {
      var {
        onChange
      } = this.props;
      var {
        sliderHeight
      } = this.state;
      var alpha = alphaForDraggerY(y, sliderHeight);
      onChange(alpha);
    };
  }

  render() {
    var {
      color,
      alpha
    } = this.props;
    var {
      sliderHeight,
      draggerHeight
    } = this.state;
    var draggerY = calculateDraggerY(alpha, sliderHeight, draggerHeight);
    var background = alphaGradientForColor(color);
    return /*#__PURE__*/React$1.createElement("div", {
      className: styles.AlphaPicker,
      ref: this.setSliderHeight
    }, /*#__PURE__*/React$1.createElement("div", {
      className: styles.ColorLayer,
      style: {
        background
      }
    }), /*#__PURE__*/React$1.createElement(Slidable$1, {
      draggerY: draggerY,
      draggerX: 0,
      onChange: this.handleChange,
      onDraggerHeight: this.setDraggerHeight
    }));
  }

}

function alphaGradientForColor(color) {
  var {
    red,
    green,
    blue
  } = hsbToRgb(color);
  var rgb = "".concat(red, ", ").concat(green, ", ").concat(blue);
  return "linear-gradient(to top, rgba(".concat(rgb, ", 0) 18px, rgba(").concat(rgb, ", 1) calc(100% - 18px))");
}

export { AlphaPicker };
