// @shopify/polaris/dist/esm/components/Page/components/Header/Header.js
import { objectSpread2 as _objectSpread2 } from '../../../../_virtual/_rollupPluginBabelHelpers.js';
import React$1, { isValidElement } from 'react';
import { useFeatures } from '../../../../utilities/features/hooks.js';
import { useMediaQuery } from '../../../../utilities/media-query/hooks.js';
import { classNames } from '../../../../utilities/css.js';
import { ConditionalWrapper, ConditionalRender } from '../../../../utilities/components.js';
import { buttonsFrom } from '../../../Button/utils.js';
import { ButtonGroup as ButtonGroup$1 } from '../../../ButtonGroup/ButtonGroup.js';
import { hasGroupsWithActions, ActionMenu as ActionMenu$1 } from '../../../ActionMenu/ActionMenu.js';
import { Breadcrumbs as Breadcrumbs$1 } from '../../../Breadcrumbs/Breadcrumbs.js';
import { Pagination as Pagination$1 } from '../../../Pagination/Pagination.js';
import { Title as Title$1 } from './components/Title/Title.js';
import styles from './Header.scss.js';

function isPrimaryAction(x) {
  return ! /*#__PURE__*/isValidElement(x) && x !== undefined;
}
function Header({
  title,
  subtitle,
  titleMetadata,
  thumbnail,
  titleHidden = false,
  separator,
  primaryAction,
  pagination,
  additionalNavigation,
  breadcrumbs = [],
  secondaryActions = [],
  actionGroups = []
}) {
  var {
    isNavigationCollapsed
  } = useMediaQuery();
  var {
    newDesignLanguage
  } = useFeatures();
  var breadcrumbMarkup = breadcrumbs.length > 0 ? /*#__PURE__*/React$1.createElement("div", {
    className: classNames(styles.BreadcrumbWrapper, newDesignLanguage && styles.newDesignLanguage)
  }, /*#__PURE__*/React$1.createElement(Breadcrumbs$1, {
    breadcrumbs: breadcrumbs
  })) : null;
  var paginationMarkup = pagination && !isNavigationCollapsed ? /*#__PURE__*/React$1.createElement("div", {
    className: styles.PaginationWrapper
  }, /*#__PURE__*/React$1.createElement(Pagination$1, Object.assign({}, pagination, {
    plain: true
  }))) : null;
  var additionalNavigationMarkup = additionalNavigation ? /*#__PURE__*/React$1.createElement("div", {
    className: styles.AdditionalNavigationWrapper
  }, additionalNavigation) : null;
  var navigationMarkup = breadcrumbMarkup || paginationMarkup || additionalNavigationMarkup ? /*#__PURE__*/React$1.createElement("div", {
    className: styles.Navigation
  }, breadcrumbMarkup, additionalNavigationMarkup, paginationMarkup) : null;
  var pageTitleMarkup = /*#__PURE__*/React$1.createElement(Title$1, {
    title: title,
    subtitle: subtitle,
    titleMetadata: titleMetadata,
    thumbnail: thumbnail
  });
  var primaryActionMarkup = primaryAction ? /*#__PURE__*/React$1.createElement(PrimaryActionMarkup, {
    primaryAction: primaryAction
  }) : null;
  var actionMenuMarkup = secondaryActions.length > 0 || hasGroupsWithActions(actionGroups) ? /*#__PURE__*/React$1.createElement(ConditionalWrapper, {
    condition: newDesignLanguage === false,
    wrapper: children => /*#__PURE__*/React$1.createElement("div", {
      className: styles.ActionMenuWrapper
    }, children)
  }, /*#__PURE__*/React$1.createElement(ActionMenu$1, {
    actions: secondaryActions,
    groups: actionGroups,
    rollup: isNavigationCollapsed
  })) : null;
  var headerClassNames = classNames(styles.Header, titleHidden && styles.titleHidden, separator && styles.separator, navigationMarkup && styles.hasNavigation, actionMenuMarkup && styles.hasActionMenu, isNavigationCollapsed && styles.mobileView, newDesignLanguage && styles.newDesignLanguage);

  if (newDesignLanguage) {
    var {
      slot1,
      slot2,
      slot3,
      slot4,
      slot5,
      slot6
    } = determineLayout({
      breadcrumbMarkup,
      pageTitleMarkup,
      paginationMarkup,
      actionMenuMarkup,
      primaryActionMarkup,
      title,
      isNavigationCollapsed
    });
    return /*#__PURE__*/React$1.createElement("div", {
      className: headerClassNames
    }, /*#__PURE__*/React$1.createElement(ConditionalRender, {
      condition: [slot1, slot2, slot3, slot4].some(notNull)
    }, /*#__PURE__*/React$1.createElement("div", {
      className: styles.Row
    }, /*#__PURE__*/React$1.createElement("div", {
      className: styles.LeftAlign
    }, slot1, slot2), /*#__PURE__*/React$1.createElement(ConditionalRender, {
      condition: [slot3, slot4].some(notNull)
    }, /*#__PURE__*/React$1.createElement("div", {
      className: styles.RightAlign
    }, /*#__PURE__*/React$1.createElement(ConditionalWrapper, {
      condition: [slot3, slot4].every(notNull),
      wrapper: children => /*#__PURE__*/React$1.createElement(ButtonGroup$1, null, children)
    }, slot3, slot4))))), /*#__PURE__*/React$1.createElement(ConditionalRender, {
      condition: [slot5, slot6].some(notNull)
    }, /*#__PURE__*/React$1.createElement("div", {
      className: styles.Row
    }, /*#__PURE__*/React$1.createElement("div", {
      className: styles.LeftAlign
    }, slot5), /*#__PURE__*/React$1.createElement(ConditionalRender, {
      condition: slot6 != null
    }, /*#__PURE__*/React$1.createElement("div", {
      className: styles.RightAlign
    }, slot6)))));
  }

  return /*#__PURE__*/React$1.createElement("div", {
    className: headerClassNames
  }, navigationMarkup, /*#__PURE__*/React$1.createElement("div", {
    className: styles.MainContent
  }, /*#__PURE__*/React$1.createElement("div", {
    className: styles.TitleActionMenuWrapper
  }, pageTitleMarkup, actionMenuMarkup), primaryActionMarkup));
}

function PrimaryActionMarkup({
  primaryAction
}) {
  var {
    isNavigationCollapsed
  } = useMediaQuery();
  var {
    newDesignLanguage
  } = useFeatures();
  var content = primaryAction;

  if (isPrimaryAction(primaryAction)) {
    var primary = primaryAction.primary === undefined ? true : primaryAction.primary;
    content = buttonsFrom(shouldShowIconOnly(newDesignLanguage, isNavigationCollapsed, primaryAction), {
      primary
    });
  }

  return /*#__PURE__*/React$1.createElement(ConditionalWrapper, {
    condition: newDesignLanguage === false,
    wrapper: children => /*#__PURE__*/React$1.createElement("div", {
      className: styles.PrimaryActionWrapper
    }, children)
  }, content);
}

function shouldShowIconOnly(newDesignLanguage, isMobile, action) {
  var {
    content,
    accessibilityLabel,
    icon
  } = action;
  if (!newDesignLanguage || icon == null) return _objectSpread2(_objectSpread2({}, action), {}, {
    icon: undefined
  });

  if (isMobile) {
    accessibilityLabel = accessibilityLabel || content;
    content = undefined;
  } else {
    icon = undefined;
  }

  return _objectSpread2(_objectSpread2({}, action), {}, {
    content,
    accessibilityLabel,
    icon
  });
}

function notNull(value) {
  return value != null;
}

function determineLayout({
  breadcrumbMarkup,
  pageTitleMarkup,
  title,
  paginationMarkup,
  actionMenuMarkup,
  primaryActionMarkup,
  isNavigationCollapsed
}) {
  var shortTitle = 20;
  var reallyShortTitle = 8; //    Header Layout
  // |----------------------------------------------------|
  // | slot1 | slot2 |                    | slot3 | slot4 |
  // |----------------------------------------------------|
  // | slot5 |                                    | slot6 |
  // |----------------------------------------------------|
  //

  var layouts = {
    mobileCompact: {
      slots: {
        slot1: null,
        slot2: pageTitleMarkup,
        slot3: actionMenuMarkup,
        slot4: primaryActionMarkup,
        slot5: null,
        slot6: null
      },
      condition: isNavigationCollapsed && breadcrumbMarkup == null && title != null && title.length <= reallyShortTitle
    },
    mobileDefault: {
      slots: {
        slot1: breadcrumbMarkup,
        slot2: null,
        slot3: actionMenuMarkup,
        slot4: primaryActionMarkup,
        slot5: pageTitleMarkup,
        slot6: null
      },
      condition: isNavigationCollapsed
    },
    desktopCompact: {
      slots: {
        slot1: breadcrumbMarkup,
        slot2: pageTitleMarkup,
        slot3: null,
        slot4: primaryActionMarkup,
        slot5: null,
        slot6: null
      },
      condition: !isNavigationCollapsed && paginationMarkup == null && actionMenuMarkup == null && title != null && title.length <= shortTitle
    },
    desktopDefault: {
      slots: {
        slot1: breadcrumbMarkup,
        slot2: pageTitleMarkup,
        slot3: null,
        slot4: paginationMarkup,
        slot5: actionMenuMarkup,
        slot6: primaryActionMarkup
      },
      condition: !isNavigationCollapsed
    }
  };
  var layout = Object.values(layouts).find(layout => layout.condition) || layouts.desktopDefault;
  return layout.slots;
}

export { Header, isPrimaryAction };
