// @shopify/polaris/dist/esm/components/ResourceList/components/BulkActions/components/BulkActionButton/BulkActionButton.js
import React$1, { useRef } from 'react';
import { Button as Button$1 } from '../../../../../Button/Button.js';
import { useComponentDidMount as useComponentDidMount$1 } from '../../../../../../utilities/use-component-did-mount.js';
import styles from '../../BulkActions.scss.js';

function BulkActionButton({
  handleMeasurement,
  url,
  external,
  onAction,
  content,
  disclosure,
  accessibilityLabel,
  disabled
}) {
  var bulkActionButton = useRef(null);
  useComponentDidMount$1(() => {
    if (handleMeasurement && bulkActionButton.current) {
      var _width = bulkActionButton.current.getBoundingClientRect().width;
      handleMeasurement(_width);
    }
  });
  return /*#__PURE__*/React$1.createElement("div", {
    className: styles.BulkActionButton,
    ref: bulkActionButton
  }, /*#__PURE__*/React$1.createElement(Button$1, {
    external: external,
    url: url,
    "aria-label": accessibilityLabel,
    onClick: onAction,
    disabled: disabled,
    disclosure: disclosure
  }, content));
}

export { BulkActionButton };
