// @shopify/polaris/dist/esm/components/RangeSlider/components/SingleThumb/SingleThumb.js
import React$1 from 'react';
import { clamp as clamp$1 } from '../../../../utilities/clamp.js';
import { classNames } from '../../../../utilities/css.js';
import { Labelled as Labelled$1, helpTextID } from '../../../Labelled/Labelled.js';
import { invertNumber as invertNumber$1 } from '../../utilities/invertNumber.js';
import { CSS_VAR_PREFIX } from '../../utilities/index.js';
import styles from './SingleThumb.scss.js';

function SingleThumb(props) {
  var {
    id,
    error,
    helpText,
    value,
    min,
    max,
    disabled,
    output,
    prefix,
    suffix,
    label,
    labelAction,
    labelHidden,
    step,
    onBlur,
    onFocus
  } = props;
  var clampedValue = clamp$1(value, min, max);
  var describedBy = [];

  if (error) {
    describedBy.push("".concat(id, "Error"));
  }

  if (helpText) {
    describedBy.push(helpTextID(id));
  }

  var ariaDescribedBy = describedBy.length ? describedBy.join(' ') : undefined;
  var sliderProgress = (clampedValue - min) * 100 / (max - min);
  var outputFactor = invertNumber$1((sliderProgress - 50) / 100);
  var cssVars = {
    ["".concat(CSS_VAR_PREFIX, "min")]: min,
    ["".concat(CSS_VAR_PREFIX, "max")]: max,
    ["".concat(CSS_VAR_PREFIX, "current")]: clampedValue,
    ["".concat(CSS_VAR_PREFIX, "progress")]: "".concat(sliderProgress, "%"),
    ["".concat(CSS_VAR_PREFIX, "output-factor")]: "".concat(outputFactor)
  };
  var outputMarkup = !disabled && output && /*#__PURE__*/React$1.createElement("output", {
    htmlFor: id,
    className: styles.Output
  }, /*#__PURE__*/React$1.createElement("div", {
    className: styles.OutputBubble
  }, /*#__PURE__*/React$1.createElement("span", {
    className: styles.OutputText
  }, clampedValue)));
  var prefixMarkup = prefix && /*#__PURE__*/React$1.createElement("div", {
    className: styles.Prefix
  }, prefix);
  var suffixMarkup = suffix && /*#__PURE__*/React$1.createElement("div", {
    className: styles.Suffix
  }, suffix);
  var className = classNames(styles.SingleThumb, error && styles.error, disabled && styles.disabled);
  return /*#__PURE__*/React$1.createElement(Labelled$1, {
    id: id,
    label: label,
    error: error,
    action: labelAction,
    labelHidden: labelHidden,
    helpText: helpText
  }, /*#__PURE__*/React$1.createElement("div", {
    className: className,
    style: cssVars
  }, prefixMarkup, /*#__PURE__*/React$1.createElement("div", {
    className: styles.InputWrapper
  }, /*#__PURE__*/React$1.createElement("input", {
    type: "range",
    className: styles.Input,
    id: id,
    name: id,
    min: min,
    max: max,
    step: step,
    value: clampedValue,
    disabled: disabled,
    onChange: handleChange,
    onFocus: onFocus,
    onBlur: onBlur,
    "aria-valuemin": min,
    "aria-valuemax": max,
    "aria-valuenow": clampedValue,
    "aria-invalid": Boolean(error),
    "aria-describedby": ariaDescribedBy
  }), outputMarkup), suffixMarkup));

  function handleChange(event) {
    var {
      onChange
    } = props;
    onChange && onChange(parseFloat(event.currentTarget.value), id);
  }
}

export { SingleThumb };
