// @shopify/polaris/dist/esm/components/RangeSlider/components/DualThumb/DualThumb.js
import { Key } from '../../../../types.js';
import React$1, { Component, createRef } from 'react';
import { FeaturesContext } from '../../../../utilities/features/context.js';
import debounce$1 from 'lodash/debounce';
import { EventListener as EventListener$1 } from '../../../EventListener/EventListener.js';
import { classNames } from '../../../../utilities/css.js';
import isEqual$1 from 'lodash/isEqual';
import { labelID } from '../../../Label/Label.js';
import { Labelled as Labelled$1 } from '../../../Labelled/Labelled.js';
import { CSS_VAR_PREFIX } from '../../utilities/index.js';
import styles from './DualThumb.scss.js';

var Control;

(function (Control) {
  Control[Control["Lower"] = 0] = "Lower";
  Control[Control["Upper"] = 1] = "Upper";
})(Control || (Control = {}));

class DualThumb extends Component {
  constructor(...args) {
    super(...args);
    this.context = void 0;
    this.state = {
      value: sanitizeValue(this.props.value, this.props.min, this.props.max, this.props.step),
      trackWidth: 0,
      trackLeft: 0
    };
    this.track = /*#__PURE__*/createRef();
    this.trackWrapper = /*#__PURE__*/createRef();
    this.thumbLower = /*#__PURE__*/createRef();
    this.thumbUpper = /*#__PURE__*/createRef();
    this.setTrackPosition = debounce$1(() => {
      if (this.track.current) {
        var newDesignLanguage = this.context && this.context.newDesignLanguage;
        var thumbSize = newDesignLanguage ? 16 : 24;
        var {
          width,
          left
        } = this.track.current.getBoundingClientRect();
        var adjustedTrackWidth = width - thumbSize;
        var adjustedTrackLeft = left + thumbSize / 2;
        this.setState({
          trackWidth: adjustedTrackWidth,
          trackLeft: adjustedTrackLeft
        });
      }
    }, 40, {
      leading: true,
      trailing: true,
      maxWait: 40
    });

    this.handleMouseDownThumbLower = event => {
      if (event.button !== 0 || this.props.disabled) return;
      registerMouseMoveHandler(this.handleMouseMoveThumbLower);
      event.stopPropagation();
    };

    this.handleMouseMoveThumbLower = event => {
      var valueUpper = this.state.value[1];
      this.setValue([this.actualXPosition(event.clientX), valueUpper], Control.Upper);
    };

    this.handleTouchStartThumbLower = event => {
      if (this.props.disabled) return;
      registerTouchMoveHandler(this.handleTouchMoveThumbLower);
      event.stopPropagation();
    };

    this.handleTouchMoveThumbLower = event => {
      event.preventDefault();
      var valueUpper = this.state.value[1];
      this.setValue([this.actualXPosition(event.touches[0].clientX), valueUpper], Control.Upper);
    };

    this.handleMouseDownThumbUpper = event => {
      if (event.button !== 0 || this.props.disabled) return;
      registerMouseMoveHandler(this.handleMouseMoveThumbUpper);
      event.stopPropagation();
    };

    this.handleMouseMoveThumbUpper = event => {
      var valueLower = this.state.value[0];
      this.setValue([valueLower, this.actualXPosition(event.clientX)], Control.Lower);
    };

    this.handleTouchStartThumbUpper = event => {
      if (this.props.disabled) return;
      registerTouchMoveHandler(this.handleTouchMoveThumbUpper);
      event.stopPropagation();
    };

    this.handleTouchMoveThumbUpper = event => {
      event.preventDefault();
      var valueLower = this.state.value[0];
      this.setValue([valueLower, this.actualXPosition(event.touches[0].clientX)], Control.Lower);
    };

    this.handleKeypressLower = event => {
      if (this.props.disabled) return;
      var {
        incrementValueLower,
        decrementValueLower
      } = this;
      var handlerMap = {
        [Key.UpArrow]: incrementValueLower,
        [Key.RightArrow]: incrementValueLower,
        [Key.DownArrow]: decrementValueLower,
        [Key.LeftArrow]: decrementValueLower
      };
      var handler = handlerMap[event.keyCode];

      if (handler != null) {
        event.preventDefault();
        event.stopPropagation();
        handler();
      }
    };

    this.handleKeypressUpper = event => {
      if (this.props.disabled) return;
      var {
        incrementValueUpper,
        decrementValueUpper
      } = this;
      var handlerMap = {
        [Key.UpArrow]: incrementValueUpper,
        [Key.RightArrow]: incrementValueUpper,
        [Key.DownArrow]: decrementValueUpper,
        [Key.LeftArrow]: decrementValueUpper
      };
      var handler = handlerMap[event.keyCode];

      if (handler != null) {
        event.preventDefault();
        event.stopPropagation();
        handler();
      }
    };

    this.incrementValueLower = () => {
      this.setValue([this.state.value[0] + this.props.step, this.state.value[1]], Control.Upper);
    };

    this.decrementValueLower = () => {
      this.setValue([this.state.value[0] - this.props.step, this.state.value[1]], Control.Upper);
    };

    this.incrementValueUpper = () => {
      this.setValue([this.state.value[0], this.state.value[1] + this.props.step], Control.Lower);
    };

    this.decrementValueUpper = () => {
      this.setValue([this.state.value[0], this.state.value[1] - this.props.step], Control.Lower);
    };

    this.dispatchValue = () => {
      var {
        onChange,
        id
      } = this.props;
      var {
        value
      } = this.state;
      onChange(value, id);
    };

    this.setValue = (dirtyValue, control) => {
      var {
        props: {
          min,
          max,
          step
        },
        state: {
          value
        }
      } = this;
      var sanitizedValue = sanitizeValue(dirtyValue, min, max, step, control);

      if (isEqual$1(sanitizedValue, value) === false) {
        this.setState({
          value: sanitizedValue
        }, this.dispatchValue);
      }
    };

    this.handleMouseDownTrack = event => {
      if (event.button !== 0 || this.props.disabled) return;
      event.preventDefault();
      var clickXPosition = this.actualXPosition(event.clientX);
      var {
        value
      } = this.state;
      var distanceFromLowerThumb = Math.abs(value[0] - clickXPosition);
      var distanceFromUpperThumb = Math.abs(value[1] - clickXPosition);

      if (distanceFromLowerThumb <= distanceFromUpperThumb) {
        this.setValue([clickXPosition, value[1]], Control.Upper);
        registerMouseMoveHandler(this.handleMouseMoveThumbLower);

        if (this.thumbLower.current != null) {
          this.thumbLower.current.focus();
        }
      } else {
        this.setValue([value[0], clickXPosition], Control.Lower);
        registerMouseMoveHandler(this.handleMouseMoveThumbUpper);

        if (this.thumbUpper.current != null) {
          this.thumbUpper.current.focus();
        }
      }
    };

    this.handleTouchStartTrack = event => {
      if (this.props.disabled) return;
      event.preventDefault();
      var clickXPosition = this.actualXPosition(event.touches[0].clientX);
      var {
        value
      } = this.state;
      var distanceFromLowerThumb = Math.abs(value[0] - clickXPosition);
      var distanceFromUpperThumb = Math.abs(value[1] - clickXPosition);

      if (distanceFromLowerThumb <= distanceFromUpperThumb) {
        this.setValue([clickXPosition, value[1]], Control.Upper);
        registerTouchMoveHandler(this.handleTouchMoveThumbLower);

        if (this.thumbLower.current != null) {
          this.thumbLower.current.focus();
        }
      } else {
        this.setValue([value[0], clickXPosition], Control.Lower);
        registerTouchMoveHandler(this.handleTouchMoveThumbUpper);

        if (this.thumbUpper.current != null) {
          this.thumbUpper.current.focus();
        }
      }
    };

    this.actualXPosition = dirtyXPosition => {
      if (this.track.current) {
        var {
          min,
          max
        } = this.props;
        var {
          trackLeft,
          trackWidth
        } = this.state;
        var relativeX = dirtyXPosition - trackLeft;
        var percentageOfTrack = relativeX / trackWidth;
        return percentageOfTrack * (max - min);
      } else {
        return 0;
      }
    };
  }

  static getDerivedStateFromProps(props, state) {
    var {
      min,
      step,
      max,
      value,
      onChange,
      id
    } = props;
    var {
      prevValue
    } = state;

    if (isEqual$1(prevValue, value)) {
      return null;
    }

    var sanitizedValue = sanitizeValue(value, min, max, step);

    if (!isEqual$1(value, sanitizedValue)) {
      onChange(sanitizedValue, id);
    }

    return {
      prevValue: value,
      value: sanitizedValue
    };
  }

  componentDidMount() {
    this.setTrackPosition();

    if (this.trackWrapper.current != null) {
      this.trackWrapper.current.addEventListener('touchstart', this.handleTouchStartTrack, {
        passive: false
      });
    }
  }

  componentWillUnmount() {
    if (this.trackWrapper.current != null) {
      this.trackWrapper.current.removeEventListener('touchstart', this.handleTouchStartTrack);
    }
  }

  render() {
    var {
      id,
      min,
      max,
      prefix,
      suffix,
      disabled,
      output,
      error,
      onFocus,
      onBlur,
      label,
      labelAction,
      labelHidden,
      helpText
    } = this.props;
    var {
      value
    } = this.state;
    var idLower = id;
    var idUpper = "".concat(id, "Upper");
    var describedBy = [];

    if (error) {
      describedBy.push("".concat(id, "Error"));
    }

    var ariaDescribedBy = describedBy.length ? describedBy.join(' ') : undefined;
    var trackWrapperClassName = classNames(styles.TrackWrapper, error && styles.error, disabled && styles.disabled);
    var thumbLowerClassName = classNames(styles.Thumbs, styles.ThumbLower, disabled && styles.disabled);
    var thumbUpperClassName = classNames(styles.Thumbs, styles.ThumbUpper, disabled && styles.disabled);
    var trackWidth = this.state.trackWidth;
    var range = max - min;
    var leftPositionThumbLower = value[0] / range * trackWidth;
    var leftPositionThumbUpper = value[1] / range * trackWidth;
    var outputLowerClassName = classNames(styles.Output, styles.OutputLower);
    var outputMarkupLower = !disabled && output ? /*#__PURE__*/React$1.createElement("output", {
      htmlFor: idLower,
      className: outputLowerClassName,
      style: {
        left: "".concat(leftPositionThumbLower, "px")
      }
    }, /*#__PURE__*/React$1.createElement("div", {
      className: styles.OutputBubble
    }, /*#__PURE__*/React$1.createElement("span", {
      className: styles.OutputText
    }, value[0]))) : null;
    var outputUpperClassName = classNames(styles.Output, styles.OutputUpper);
    var outputMarkupUpper = !disabled && output ? /*#__PURE__*/React$1.createElement("output", {
      htmlFor: idUpper,
      className: outputUpperClassName,
      style: {
        left: "".concat(leftPositionThumbUpper, "px")
      }
    }, /*#__PURE__*/React$1.createElement("div", {
      className: styles.OutputBubble
    }, /*#__PURE__*/React$1.createElement("span", {
      className: styles.OutputText
    }, value[1]))) : null;
    var cssVars = {
      ["".concat(CSS_VAR_PREFIX, "progress-lower")]: "".concat(leftPositionThumbLower, "px"),
      ["".concat(CSS_VAR_PREFIX, "progress-upper")]: "".concat(leftPositionThumbUpper, "px")
    };
    var prefixMarkup = prefix && /*#__PURE__*/React$1.createElement("div", {
      className: styles.Prefix
    }, prefix);
    var suffixMarkup = suffix && /*#__PURE__*/React$1.createElement("div", {
      className: styles.Suffix
    }, suffix);
    return /*#__PURE__*/React$1.createElement(React$1.Fragment, null, /*#__PURE__*/React$1.createElement(Labelled$1, {
      id: id,
      label: label,
      error: error,
      action: labelAction,
      labelHidden: labelHidden,
      helpText: helpText
    }, /*#__PURE__*/React$1.createElement("div", {
      className: styles.Wrapper
    }, prefixMarkup, /*#__PURE__*/React$1.createElement("div", {
      className: trackWrapperClassName,
      onMouseDown: this.handleMouseDownTrack,
      ref: this.trackWrapper
    }, /*#__PURE__*/React$1.createElement("div", {
      className: styles.Track,
      style: cssVars,
      ref: this.track
    }), /*#__PURE__*/React$1.createElement("div", {
      className: styles['Track--dashed']
    }), /*#__PURE__*/React$1.createElement("button", {
      id: idLower,
      className: thumbLowerClassName,
      style: {
        left: "".concat(leftPositionThumbLower, "px")
      },
      role: "slider",
      "aria-disabled": disabled,
      "aria-valuemin": min,
      "aria-valuemax": max,
      "aria-valuenow": value[0],
      "aria-invalid": Boolean(error),
      "aria-describedby": ariaDescribedBy,
      "aria-labelledby": labelID(id),
      onFocus: onFocus,
      onBlur: onBlur,
      onKeyDown: this.handleKeypressLower,
      onMouseDown: this.handleMouseDownThumbLower,
      onTouchStart: this.handleTouchStartThumbLower,
      ref: this.thumbLower,
      disabled: disabled
    }), outputMarkupLower, /*#__PURE__*/React$1.createElement("button", {
      id: idUpper,
      className: thumbUpperClassName,
      style: {
        left: "".concat(leftPositionThumbUpper, "px")
      },
      role: "slider",
      "aria-disabled": disabled,
      "aria-valuemin": min,
      "aria-valuemax": max,
      "aria-valuenow": value[1],
      "aria-invalid": Boolean(error),
      "aria-describedby": ariaDescribedBy,
      "aria-labelledby": labelID(id),
      onFocus: onFocus,
      onBlur: onBlur,
      onKeyDown: this.handleKeypressUpper,
      onMouseDown: this.handleMouseDownThumbUpper,
      onTouchStart: this.handleTouchStartThumbUpper,
      ref: this.thumbUpper,
      disabled: disabled
    }), outputMarkupUpper), suffixMarkup)), /*#__PURE__*/React$1.createElement(EventListener$1, {
      event: "resize",
      handler: this.setTrackPosition
    }));
  }

}
DualThumb.contextType = FeaturesContext;

function registerMouseMoveHandler(handler) {
  document.addEventListener('mousemove', handler);
  document.addEventListener('mouseup', () => {
    document.removeEventListener('mousemove', handler);
  }, {
    once: true
  });
}

function registerTouchMoveHandler(handler) {
  var removeHandler = () => {
    document.removeEventListener('touchmove', handler);
    document.removeEventListener('touchend', removeHandler);
    document.removeEventListener('touchcancel', removeHandler);
  };

  document.addEventListener('touchmove', handler, {
    passive: false
  });
  document.addEventListener('touchend', removeHandler, {
    once: true
  });
  document.addEventListener('touchcancel', removeHandler, {
    once: true
  });
}

function sanitizeValue(value, min, max, step, control = Control.Upper) {
  var upperValue = inBoundsUpper(roundedToStep(value[1]));
  var lowerValue = inBoundsLower(roundedToStep(value[0]));
  var maxLowerValue = upperValue - step;
  var minUpperValue = lowerValue + step;

  if (control === Control.Upper && lowerValue > maxLowerValue) {
    lowerValue = maxLowerValue;
  } else if (control === Control.Lower && upperValue < minUpperValue) {
    upperValue = minUpperValue;
  }

  return [lowerValue, upperValue];

  function inBoundsUpper(value) {
    var lowerMin = min + step;

    if (value < lowerMin) {
      return lowerMin;
    } else if (value > max) {
      return max;
    } else {
      return value;
    }
  }

  function inBoundsLower(value) {
    var upperMax = max - step;

    if (value < min) {
      return min;
    } else if (value > upperMax) {
      return upperMax;
    } else {
      return value;
    }
  }

  function roundedToStep(value) {
    return Math.round(value / step) * step;
  }
}

export { DualThumb };
