// @shopify/polaris/dist/esm/components/Tabs/components/Item/Item.js
import React$1, { PureComponent } from 'react';
import { FeaturesContext } from '../../../../utilities/features/context.js';
import { classNames } from '../../../../utilities/css.js';
import { UnstyledLink as UnstyledLink$1 } from '../../../UnstyledLink/UnstyledLink.js';
import styles from '../../Tabs.scss.js';

class Item extends PureComponent {
  constructor(...args) {
    super(...args);
    this.context = void 0;
    this.focusedNode = null;

    this.setFocusedNode = node => {
      this.focusedNode = node;
    };
  }

  componentDidMount() {
    var {
      focusedNode
    } = this;
    var {
      focused
    } = this.props;

    if (focusedNode && focusedNode instanceof HTMLElement && focused) {
      focusedNode.focus();
    }
  }

  componentDidUpdate() {
    var {
      focusedNode
    } = this;
    var {
      focused
    } = this.props;

    if (focusedNode && focusedNode instanceof HTMLElement && focused) {
      focusedNode.focus();
    }
  }

  render() {
    var {
      id,
      panelID,
      children,
      url,
      accessibilityLabel,
      onClick = noop
    } = this.props;
    var {
      newDesignLanguage
    } = this.context || {};
    var classname = classNames(styles.Item, newDesignLanguage && styles.newDesignLanguage);
    var sharedProps = {
      id,
      ref: this.setFocusedNode,
      onClick,
      className: classname,
      'aria-controls': panelID,
      'aria-selected': false,
      'aria-label': accessibilityLabel
    };
    var markup = url ? /*#__PURE__*/React$1.createElement(UnstyledLink$1, Object.assign({}, sharedProps, {
      url: url
    }), children) : /*#__PURE__*/React$1.createElement("button", Object.assign({}, sharedProps, {
      type: "button"
    }), children);
    return /*#__PURE__*/React$1.createElement("li", null, markup);
  }

}
Item.contextType = FeaturesContext;

function noop() {}

export { Item };
