// @shopify/polaris/dist/esm/components/Navigation/components/Section/Section.js
import { objectWithoutProperties as _objectWithoutProperties } from '../../../../_virtual/_rollupPluginBabelHelpers.js';
import React$1, { useRef, useEffect } from 'react';
import { navigationBarCollapsed } from '../../../../utilities/breakpoints.js';
import { useUniqueId } from '../../../../utilities/unique-id/hooks.js';
import { classNames } from '../../../../utilities/css.js';
import { HorizontalDotsMinor } from '@shopify/polaris-icons';
import { Icon as Icon$1 } from '../../../Icon/Icon.js';
import { useToggle as useToggle$1 } from '../../../../utilities/use-toggle.js';
import { Collapsible as Collapsible$1 } from '../../../Collapsible/Collapsible.js';
import styles from '../../Navigation.scss.js';
import { Item as Item$1 } from '../Item/Item.js';

var _ref = /*#__PURE__*/React$1.createElement(Icon$1, {
  source: HorizontalDotsMinor
});

function Section({
  title,
  fill,
  action,
  items,
  rollup,
  separator
}) {
  var {
    value: expanded,
    toggle: toggleExpanded,
    setFalse: setExpandedFalse
  } = useToggle$1(false);
  var animationFrame = useRef(null);

  var handleClick = (onClick, hasSubNavItems) => {
    return () => {
      if (onClick) {
        onClick();
      }

      if (animationFrame.current) {
        cancelAnimationFrame(animationFrame.current);
      }

      if (!hasSubNavItems || !navigationBarCollapsed().matches) {
        animationFrame.current = requestAnimationFrame(setExpandedFalse);
      }
    };
  };

  useEffect(() => {
    return () => {
      animationFrame.current && cancelAnimationFrame(animationFrame.current);
    };
  });
  var className = classNames(styles.Section, separator && styles['Section-withSeparator'], fill && styles['Section-fill']);
  var actionMarkup = action && /*#__PURE__*/React$1.createElement("button", {
    type: "button",
    className: styles.Action,
    "aria-label": action.accessibilityLabel,
    onClick: action.onClick
  }, /*#__PURE__*/React$1.createElement(Icon$1, {
    source: action.icon
  }));
  var sectionHeadingMarkup = title && /*#__PURE__*/React$1.createElement("li", {
    className: styles.SectionHeading
  }, /*#__PURE__*/React$1.createElement("span", {
    className: styles.Text
  }, title), actionMarkup);
  var itemsMarkup = items.map(item => {
    var {
      onClick,
      label,
      subNavigationItems
    } = item,
        rest = _objectWithoutProperties(item, ["onClick", "label", "subNavigationItems"]);

    var hasSubNavItems = subNavigationItems != null && subNavigationItems.length > 0;
    return /*#__PURE__*/React$1.createElement(Item$1, Object.assign({}, rest, {
      key: label,
      label: label,
      subNavigationItems: subNavigationItems,
      onClick: handleClick(onClick, hasSubNavItems)
    }));
  });
  var toggleClassName = classNames(styles.Item, styles.RollupToggle);
  var ariaLabel = rollup && (expanded ? rollup.hide : rollup.view);
  var toggleRollup = rollup && items.length > rollup.after && /*#__PURE__*/React$1.createElement("div", {
    className: styles.ListItem,
    key: "List Item"
  }, /*#__PURE__*/React$1.createElement("button", {
    type: "button",
    className: toggleClassName,
    onClick: toggleExpanded,
    "aria-label": ariaLabel
  }, /*#__PURE__*/React$1.createElement("span", {
    className: styles.Icon
  }, _ref)));
  var activeItemIndex = items.findIndex(item => {
    if (!rollup) {
      return false;
    }

    return rollup.activePath === item.url || item.url && rollup.activePath.startsWith(item.url) || (item.subNavigationItems ? item.subNavigationItems.some(({
      url: itemUrl
    }) => rollup.activePath.startsWith(itemUrl)) : false);
  });
  var sectionItems = rollup ? itemsMarkup.slice(0, rollup.after) : itemsMarkup;
  var additionalItems = rollup ? itemsMarkup.slice(rollup.after) : [];

  if (rollup && activeItemIndex !== -1 && activeItemIndex > rollup.after - 1) {
    sectionItems.push(...additionalItems.splice(activeItemIndex - rollup.after, 1));
  }

  var additionalItemsId = useUniqueId('AdditionalItems');
  var activeItemsMarkup = rollup && additionalItems.length > 0 && /*#__PURE__*/React$1.createElement("li", {
    className: styles.RollupSection
  }, /*#__PURE__*/React$1.createElement(Collapsible$1, {
    id: additionalItemsId,
    open: expanded
  }, /*#__PURE__*/React$1.createElement("ul", {
    className: styles.List
  }, additionalItems)), toggleRollup);
  return /*#__PURE__*/React$1.createElement("ul", {
    className: className
  }, sectionHeadingMarkup, sectionItems, activeItemsMarkup);
}

export { Section };
