// @shopify/polaris/dist/esm/components/DatePicker/components/Day/Day.js
import React$1, { memo, useRef, useEffect } from 'react';
import { useI18n } from '../../../../utilities/i18n/hooks.js';
import { classNames } from '../../../../utilities/css.js';
import { isSameDay } from '../../../../utilities/dates.js';
import { monthName } from '../../utilities.js';
import styles from '../../DatePicker.scss.js';

var Day = /*#__PURE__*/memo(function Day({
  day,
  focused,
  onClick,
  onHover = noop,
  onFocus = noop,
  selected,
  inRange,
  inHoveringRange,
  disabled,
  lastDayOfMonth,
  isLastSelectedDay,
  isFirstSelectedDay,
  isHoveringRight,
  rangeIsDifferent
}) {
  var i18n = useI18n();
  var dayNode = useRef(null);
  var hoverValue = lastDayOfMonth || day;
  useEffect(() => {
    if (focused && dayNode.current) {
      dayNode.current.focus();
    }
  }, [focused]);

  if (!day) {
    return /*#__PURE__*/React$1.createElement("div", {
      className: styles.EmptyDay,
      onMouseOver: () => onHover(hoverValue)
    });
  }

  var handleClick = onClick && !disabled ? onClick.bind(null, day) : noop;
  var today = isSameDay(new Date(), day);
  var className = classNames(styles.Day, selected && styles['Day-selected'], disabled && styles['Day-disabled'], today && styles['Day-today'], (inRange || inHoveringRange) && !disabled && styles['Day-inRange'], isLastSelectedDay && styles['Day-lastInRange'], isFirstSelectedDay && styles['Day-firstInRange'], isHoveringRight && styles['Day-hoverRight'], rangeIsDifferent && styles['Day-hasRange']);
  var date = day.getDate();
  var tabIndex = (focused || selected || today || date === 1) && !disabled ? 0 : -1;
  var ariaLabel = ["".concat(today ? i18n.translate('Polaris.DatePicker.today') : ''), "".concat(i18n.translate("Polaris.DatePicker.months.".concat(monthName(day.getMonth()))), " "), "".concat(date, " "), "".concat(day.getFullYear())].join('');
  return /*#__PURE__*/React$1.createElement("button", {
    onFocus: () => onFocus(day),
    type: "button",
    ref: dayNode,
    tabIndex: tabIndex,
    className: className,
    onMouseOver: () => onHover(hoverValue),
    onClick: handleClick,
    "aria-label": ariaLabel,
    "aria-selected": selected,
    "aria-disabled": disabled,
    role: "gridcell"
  }, date);
});

function noop() {}

export { Day };
