// @shopify/polaris/dist/esm/components/DataTable/components/Navigation/Navigation.js
import React$1 from 'react';
import { useI18n } from '../../../../utilities/i18n/hooks.js';
import { classNames } from '../../../../utilities/css.js';
import { ChevronLeftMinor, ChevronRightMinor } from '@shopify/polaris-icons';
import { Button as Button$1 } from '../../../Button/Button.js';
import styles from '../../DataTable.scss.js';

function Navigation({
  columnVisibilityData,
  isScrolledFarthestLeft,
  isScrolledFarthestRight,
  navigateTableLeft,
  navigateTableRight
}) {
  var i18n = useI18n();
  var pipMarkup = columnVisibilityData.map((column, index) => {
    var className = classNames(styles.Pip, column.isVisible && styles['Pip-visible']);
    return /*#__PURE__*/React$1.createElement("div", {
      className: className,
      key: "pip-".concat(index)
    });
  });
  var leftA11yLabel = i18n.translate('Polaris.DataTable.navAccessibilityLabel', {
    direction: 'left'
  });
  var rightA11yLabel = i18n.translate('Polaris.DataTable.navAccessibilityLabel', {
    direction: 'right'
  });
  return /*#__PURE__*/React$1.createElement("div", {
    className: styles.Navigation
  }, /*#__PURE__*/React$1.createElement(Button$1, {
    plain: true,
    icon: ChevronLeftMinor,
    disabled: isScrolledFarthestLeft,
    accessibilityLabel: leftA11yLabel,
    onClick: navigateTableLeft
  }), pipMarkup, /*#__PURE__*/React$1.createElement(Button$1, {
    plain: true,
    icon: ChevronRightMinor,
    disabled: isScrolledFarthestRight,
    accessibilityLabel: rightA11yLabel,
    onClick: navigateTableRight
  }));
}

export { Navigation };
