// @shopify/polaris/dist/esm/components/Frame/components/Toast/Toast.js
import { Key } from '../../../../types.js';
import React$1, { useEffect } from 'react';
import { ThemeProvider as ThemeProvider$1 } from '../../../ThemeProvider/ThemeProvider.js';
import { classNames } from '../../../../utilities/css.js';
import { MobileCancelMajorMonotone } from '@shopify/polaris-icons';
import { Icon as Icon$1 } from '../../../Icon/Icon.js';
import { KeypressListener as KeypressListener$1 } from '../../../KeypressListener/KeypressListener.js';
import { Button as Button$1 } from '../../../Button/Button.js';
import styles from './Toast.scss.js';

var DEFAULT_TOAST_DURATION = 5000;
var DEFAULT_TOAST_DURATION_WITH_ACTION = 10000;

var _ref = /*#__PURE__*/React$1.createElement(Icon$1, {
  source: MobileCancelMajorMonotone
});

function Toast({
  content,
  onDismiss,
  duration,
  error,
  action
}) {
  useEffect(() => {
    var timeoutDuration = duration || DEFAULT_TOAST_DURATION;

    if (action && !duration) {
      timeoutDuration = DEFAULT_TOAST_DURATION_WITH_ACTION;
    } else if (action && duration && duration < DEFAULT_TOAST_DURATION_WITH_ACTION) {
      // eslint-disable-next-line no-console
      console.log('Toast with action should persist for at least 10,000 milliseconds to give the merchant enough time to act on it.');
    }

    var timer = setTimeout(onDismiss, timeoutDuration);
    return () => {
      clearTimeout(timer);
    };
  }, [action, duration, onDismiss]);
  var dismissMarkup = /*#__PURE__*/React$1.createElement("button", {
    type: "button",
    className: styles.CloseButton,
    onClick: onDismiss
  }, _ref);
  var actionMarkup = action ? /*#__PURE__*/React$1.createElement("div", {
    className: styles.Action
  }, /*#__PURE__*/React$1.createElement(Button$1, {
    plain: true,
    monochrome: true,
    onClick: action.onAction
  }, action.content)) : null;
  var className = classNames(styles.Toast, error && styles.error);
  return /*#__PURE__*/React$1.createElement(ThemeProvider$1, {
    theme: {
      colorScheme: 'inverse'
    }
  }, /*#__PURE__*/React$1.createElement("div", {
    className: className
  }, /*#__PURE__*/React$1.createElement(KeypressListener$1, {
    keyCode: Key.Escape,
    handler: onDismiss
  }), content, actionMarkup, dismissMarkup));
}

export { DEFAULT_TOAST_DURATION, DEFAULT_TOAST_DURATION_WITH_ACTION, Toast };
